x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -0,0 +1,216 @@
+        PARSER.declareStringArray(constructorArg(), IDS_FIELD);
+        declareStandardFields(PARSER);
+    }
+
+    public static PinnedQueryBuilder fromXContent(XContentParser parser) {
+        try {
+            return PARSER.apply(parser, null);
+        } catch (IllegalArgumentException e) {
+            throw new ParsingException(parser.getTokenLocation(), e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryShardContext) throws IOException {
+        if (ids.size() == 0 && organicQuery == null) {
+            return new MatchNoneQueryBuilder();
+        }
+        return this;
+    }
+
+    @Override
+    protected Query doToQuery(QueryShardContext context) throws IOException {
+        MappedFieldType idField = context.fieldMapper(IdFieldMapper.NAME);
+        if (idField == null) {
+            return new MatchNoDocsQuery("No mappings");
+        }
+        if (this.ids.isEmpty()) {
+            return new CappedScoreQuery(organicQuery.toQuery(context), MAX_ORGANIC_SCORE);
+        } else {
+            BooleanQuery.Builder pinnedQueries = new BooleanQuery.Builder();
+
+            // Ensure each pin order using a Boost query with the relevant boost factor
+            int minPin = NumericUtils.floatToSortableInt(MAX_ORGANIC_SCORE) + 1;
+            int boostNum = minPin + ids.size();
+            float lastScore = Float.MAX_VALUE;
+            for (String id : ids) {
+                float pinScore = NumericUtils.sortableIntToFloat(boostNum);
+                assert pinScore < lastScore;
+                lastScore = pinScore;
+                boostNum--;
+                // Ensure the pin order using a Boost query with the relevant boost factor
+                Query idQuery = new BoostQuery(new ConstantScoreQuery(idField.termQuery(id, context)), pinScore);
+                pinnedQueries.add(idQuery, BooleanClause.Occur.SHOULD);
+            }
+
+            // Score for any pinned query clause should be used, regardless of any organic clause score, to preserve pin order.
+            // Use dismax to always take the larger (ie pinned) of the organic vs pinned scores
+            List<Query> organicAndPinned = new ArrayList<>();
+            organicAndPinned.add(pinnedQueries.build());
+            // Cap the scores of the organic query
+            organicAndPinned.add(new CappedScoreQuery(organicQuery.toQuery(context), MAX_ORGANIC_SCORE));
+            return new DisjunctionMaxQuery(organicAndPinned, 0);
+        }
+
+    }
+
+    @Override
+    protected int doHashCode() {
+        return Objects.hash(ids, organicQuery);
+    }
+
+    @Override
+    protected boolean doEquals(PinnedQueryBuilder other) {
+        return Objects.equals(ids, other.ids) && Objects.equals(organicQuery, other.organicQuery);
```

Comment:
I think this is problematic because two linked hash sets are equal if they have the same content, even if it is out of order. So maybe we need to create a set in the constructor for validation purposes, but store the list of ids internally as a List?